File: ext/go/pxb/fields.go
package pxb

import (
	"encoding/binary"
	"errors"
)

// Tagged payload wire kinds. Only these two exist so unknown fields are
// always skippable without a schema.
const (
	WireU64   uint8 = 1 // 8-byte little-endian (u16/u32/bool/event codes)
	WireBytes uint8 = 2 // u32 length + bytes
)

var (
	ErrBadWire = errors.New("pxb: bad wire kind")
	ErrBadTag  = errors.New("pxb: bad field tag")
)

// FieldWriter builds a tagged-field payload (protobuf-style, fixed-width).
//
// Layout per field:
//
//	tag u16 | kind u8 | value…
//
// Unknown tags are skipped by kind. New fields must use new tag numbers;
// never reuse or reorder meaning of an existing tag.
type FieldWriter struct {
	b []byte
}

// Reset clears the buffer keeping capacity.
func (fw *FieldWriter) Reset() { fw.b = fw.b[:0] }

// Bytes returns the encoded payload.
func (fw *FieldWriter) Bytes() []byte { return fw.b }

func (fw *FieldWriter) putHdr(tag uint16, kind uint8) {
	var tmp [3]byte
	binary.LittleEndian.PutUint16(tmp[0:2], tag)
	tmp[2] = kind
	fw.b = append(fw.b, tmp[:]...)
}

// PutU64 writes an 8-byte integer field.
func (fw *FieldWriter) PutU64(tag uint16, v uint64) {
	fw.putHdr(tag, WireU64)
	var tmp [8]byte
	binary.LittleEndian.PutUint64(tmp[:], v)
	fw.b = append(fw.b, tmp[:]...)
}

// PutU16 writes v as WireU64.
func (fw *FieldWriter) PutU16(tag, v uint16) { fw.PutU64(tag, uint64(v)) }

// PutU32 writes v as WireU64.
func (fw *FieldWriter) PutU32(tag uint16, v uint32) { fw.PutU64(tag, uint64(v)) }

// PutBool writes v as WireU64 0/1.
func (fw *FieldWriter) PutBool(tag uint16, v bool) {
	if v {
		fw.PutU64(tag, 1)
	} else {
		fw.PutU64(tag, 0)
	}
}

// PutBytes writes a length-prefixed blob. Empty blobs are omitted so
// decoders stay compact on the hot path.
func (fw *FieldWriter) PutBytes(tag uint16, p []byte) {
	if len(p) == 0 {
		return
	}
	fw.putHdr(tag, WireBytes)
	n := uint32(len(p)) //nolint:gosec // G115: blob length bounded by MaxPayload on the frame path
	var tmp [4]byte
	binary.LittleEndian.PutUint32(tmp[:], n)
	fw.b = append(fw.b, tmp[:]...)
	fw.b = append(fw.b, p...)
}

// PutString writes s as WireBytes. Empty strings are omitted.
func (fw *FieldWriter) PutString(tag uint16, s string) {
	if s == "" {
		return
	}
	fw.PutBytes(tag, []byte(s))
}

// PutU16s writes a packed list as WireBytes: u16 count + u16 values.
func (fw *FieldWriter) PutU16s(tag uint16, vs []uint16) {
	if len(vs) == 0 {
		return
	}
	var inner ByteWriter
	inner.U16s(vs)
	fw.PutBytes(tag, inner.Bytes())
}

// FieldReader walks a tagged-field payload.
type FieldReader struct {
	b []byte
	i int
}

// NewFieldReader wraps b.
func NewFieldReader(b []byte) *FieldReader { return &FieldReader{b: b} }

// Done reports whether all bytes were consumed.
func (fr *FieldReader) Done() bool { return fr.i >= len(fr.b) }

func (fr *FieldReader) need(n int) error {
	if len(fr.b)-fr.i < n {
		return ErrTruncated
	}
	return nil
}

// Next returns the next field tag and wire kind.
func (fr *FieldReader) Next() (tag uint16, kind uint8, err error) {
	if fr.Done() {
		return 0, 0, ErrTruncated
	}
	if err := fr.need(3); err != nil {
		return 0, 0, err
	}
	tag = uint16(fr.b[fr.i]) | uint16(fr.b[fr.i+1])<<8
	kind = fr.b[fr.i+2]
	fr.i += 3
	if kind != WireU64 && kind != WireBytes {
		return tag, kind, ErrBadWire
	}
	if tag == 0 {
		return tag, kind, ErrBadTag
	}
	return tag, kind, nil
}

// U64 reads a WireU64 value (call after Next returned WireU64).
func (fr *FieldReader) U64() (uint64, error) {
	if err := fr.need(8); err != nil {
		return 0, err
	}
	b := fr.b[fr.i:]
	fr.i += 8
	return uint64(b[0]) | uint64(b[1])<<8 | uint64(b[2])<<16 | uint64(b[3])<<24 |
		uint64(b[4])<<32 | uint64(b[5])<<40 | uint64(b[6])<<48 | uint64(b[7])<<56, nil
}

// Bytes reads a WireBytes value (call after Next returned WireBytes).
// The returned slice aliases the underlying buffer.
func (fr *FieldReader) Bytes() ([]byte, error) {
	if err := fr.need(4); err != nil {
		return nil, err
	}
	n := uint32(fr.b[fr.i]) | uint32(fr.b[fr.i+1])<<8 | uint32(fr.b[fr.i+2])<<16 | uint32(fr.b[fr.i+3])<<24
	fr.i += 4
	if err := fr.need(int(n)); err != nil {
		return nil, err
	}
	out := fr.b[fr.i : fr.i+int(n)]
	fr.i += int(n)
	return out, nil
}

// Skip discards the value for kind.
func (fr *FieldReader) Skip(kind uint8) error {
	switch kind {
	case WireU64:
		_, err := fr.U64()
		return err
	case WireBytes:
		_, err := fr.Bytes()
		return err
	default:
		return ErrBadWire
	}
}

// Walk calls fn for each field. Unknown tags should Skip via the reader.
// fn may call U64/Bytes exactly once for the current field, or Skip.
func Walk(b []byte, fn func(tag uint16, kind uint8, fr *FieldReader) error) error {
	fr := NewFieldReader(b)
	for !fr.Done() {
		tag, kind, err := fr.Next()
		if err != nil {
			return err
		}
		before := fr.i
		if err := fn(tag, kind, fr); err != nil {
			return err
		}
		// If the callback neither consumed nor skipped, skip for them.
		if fr.i == before {
			if err := fr.Skip(kind); err != nil {
				return err
			}
		}
	}
	return nil
}

File: ext/rust/src/pxb/fields.rs
//! Tagged-field payloads (protobuf-style, fixed-width).
//!
//! Layout per field: `tag u16 | kind u8 | value…`. Only two wire kinds exist,
//! so unknown fields are always skippable without a schema. Decoders must skip
//! unknown tags; omitting a tag means the zero value.
//!
//! Evolution rules (see `doc/extensions.md`):
//! - new fields take a new tag number, never reuse a tag;
//! - empty strings/blobs are omitted on encode;
//! - experimental tags use 128+ and must remain skippable.

use super::codec::{Error, Error::*};

/// 8-byte little-endian (u16/u32/bool/event codes share this kind).
pub const WIRE_U64: u8 = 1;
/// u32 length + bytes.
pub const WIRE_BYTES: u8 = 2;

/// Builds a tagged-field payload.
pub struct FieldWriter {
    buf: Vec<u8>,
}

impl Default for FieldWriter {
    fn default() -> Self {
        Self::new()
    }
}

impl FieldWriter {
    pub fn new() -> Self {
        Self {
            buf: Vec::with_capacity(32),
        }
    }

    /// Returns the encoded payload.
    pub fn bytes(&self) -> &[u8] {
        &self.buf
    }

    pub fn into_vec(self) -> Vec<u8> {
        self.buf
    }

    fn put_hdr(&mut self, tag: u16, kind: u8) {
        self.buf.extend_from_slice(&tag.to_le_bytes());
        self.buf.push(kind);
    }

    /// Writes an 8-byte integer field. Always written, even when zero —
    /// decoders rely on the tag being present for bool-ish fields.
    pub fn put_u64(&mut self, tag: u16, v: u64) {
        self.put_hdr(tag, WIRE_U64);
        self.buf.extend_from_slice(&v.to_le_bytes());
    }

    pub fn put_u16(&mut self, tag: u16, v: u16) {
        self.put_u64(tag, u64::from(v));
    }

    pub fn put_u32(&mut self, tag: u16, v: u32) {
        self.put_u64(tag, u64::from(v));
    }

    pub fn put_bool(&mut self, tag: u16, v: bool) {
        self.put_u64(tag, u64::from(v));
    }

    /// Writes a length-prefixed blob. Empty blobs are omitted so decoders stay
    /// compact on the hot path.
    pub fn put_bytes(&mut self, tag: u16, p: &[u8]) {
        if p.is_empty() {
            return;
        }
        self.put_hdr(tag, WIRE_BYTES);
        self.buf.extend_from_slice(&(p.len() as u32).to_le_bytes());
        self.buf.extend_from_slice(p);
    }

    pub fn put_string(&mut self, tag: u16, s: &str) {
        if s.is_empty() {
            return;
        }
        self.put_bytes(tag, s.as_bytes());
    }

    /// Writes a packed list as `WireBytes`: `u16 count` + values.
    pub fn put_u16s(&mut self, tag: u16, vs: &[u16]) {
        if vs.is_empty() {
            return;
        }
        let mut inner = Vec::with_capacity(2 + vs.len() * 2);
        inner.extend_from_slice(&(vs.len() as u16).to_le_bytes());
        inner.extend(vs.iter().flat_map(|v| v.to_le_bytes()));
        self.put_bytes(tag, &inner);
    }
}

/// Walks a tagged-field payload.
pub struct FieldReader<'a> {
    b: &'a [u8],
    i: usize,
}

impl<'a> FieldReader<'a> {
    pub fn new(b: &'a [u8]) -> Self {
        Self { b, i: 0 }
    }

    pub fn done(&self) -> bool {
        self.i >= self.b.len()
    }

    fn need(&self, n: usize) -> Result<(), Error> {
        if self.b.len() - self.i < n {
            return Err(Truncated);
        }
        Ok(())
    }

    /// Returns the next field tag and wire kind.
    pub fn next_field(&mut self) -> Result<(u16, u8), Error> {
        if self.done() {
            return Err(Truncated);
        }
        self.need(3)?;
        let tag = u16::from_le_bytes([self.b[self.i], self.b[self.i + 1]]);
        let kind = self.b[self.i + 2];
        self.i += 3;
        if kind != WIRE_U64 && kind != WIRE_BYTES {
            return Err(BadWire);
        }
        if tag == 0 {
            return Err(BadTag);
        }
        Ok((tag, kind))
    }

    /// Reads a `WireU64` value (call after `next` returned `WireU64`).
    pub fn u64(&mut self) -> Result<u64, Error> {
        self.need(8)?;
        let v = u64::from_le_bytes(self.b[self.i..self.i + 8].try_into().unwrap());
        self.i += 8;
        Ok(v)
    }

    /// Reads a `WireBytes` value (call after `next` returned `WireBytes`).
    /// The returned slice aliases the payload.
    pub fn bytes(&mut self) -> Result<&'a [u8], Error> {
        self.need(4)?;
        let n = u32::from_le_bytes(self.b[self.i..self.i + 4].try_into().unwrap()) as usize;
        self.i += 4;
        self.need(n)?;
        let out = &self.b[self.i..self.i + n];
        self.i += n;
        Ok(out)
    }

    /// Discards the value for `kind`.
    pub fn skip(&mut self, kind: u8) -> Result<(), Error> {
        match kind {
            WIRE_U64 => {
                self.u64()?;
                Ok(())
            }
            WIRE_BYTES => {
                self.bytes()?;
                Ok(())
            }
            _ => Err(BadWire),
        }
    }
}

/// Calls `f` for each field; unknown tags should be skipped via the reader.
/// `f` may call `u64`/`bytes` exactly once for the current field, or `skip`.
pub fn walk_fields(
    b: &[u8],
    mut f: impl FnMut(u16, u8, &mut FieldReader<'_>) -> Result<(), Error>,
) -> Result<(), Error> {
    let mut fr = FieldReader::new(b);
    while !fr.done() {
        let (tag, kind) = fr.next_field()?;
        let before = fr.i;
        f(tag, kind, &mut fr)?;
        // If the callback neither consumed nor skipped, skip for them.
        if fr.i == before {
            fr.skip(kind)?;
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn writer_reader_roundtrip() {
        let mut w = FieldWriter::new();
        w.put_u64(1, 42);
        w.put_u16(2, 7);
        w.put_bool(3, true);
        w.put_bool(4, false);
        w.put_string(5, "hello");
        w.put_bytes(6, &[1, 2, 3]);
        w.put_u16s(7, &[5, 10]);

        let mut seen = Vec::new();
        walk_fields(w.bytes(), |tag, kind, fr| {
            let v = match kind {
                WIRE_U64 => fr.u64().unwrap(),
                WIRE_BYTES => {
                    let b = fr.bytes().unwrap();
                    b.len() as u64
                }
                _ => unreachable!(),
            };
            seen.push((tag, v));
            Ok(())
        })
        .unwrap();
        assert_eq!(
            seen,
            vec![(1, 42), (2, 7), (3, 1), (4, 0), (5, 5), (6, 3), (7, 6)]
        );
    }

    #[test]
    fn empty_values_are_omitted() {
        let mut w = FieldWriter::new();
        w.put_string(1, "");
        w.put_bytes(2, &[]);
        w.put_u16s(3, &[]);
        assert!(w.bytes().is_empty());
    }

    #[test]
    fn unknown_tags_are_skippable() {
        let mut w = FieldWriter::new();
        w.put_u64(1, 9);
        w.put_string(128, "experimental");
        w.put_string(2, "known");

        let mut name = String::new();
        walk_fields(w.bytes(), |tag, kind, fr| {
            match tag {
                2 => name = String::from_utf8_lossy(fr.bytes().unwrap()).into_owned(),
                _ => fr.skip(kind)?,
            }
            Ok(())
        })
        .unwrap();
        assert_eq!(name, "known");
    }

    #[test]
    fn rejects_bad_wire_kinds() {
        let mut w = FieldWriter::new();
        w.buf = vec![1, 0, 7]; // tag 1, unknown kind 7, no value
        let err = walk_fields(w.bytes(), |_t, _k, _fr| Ok(())).unwrap_err();
        assert!(matches!(err, Error::BadWire));
    }

    #[test]
    fn truncated_payload_errors() {
        let mut w = FieldWriter::new();
        w.put_u64(1, 42);
        let bytes = w.into_vec();
        let err = walk_fields(&bytes[..3], |_t, _k, _fr| Ok(())).unwrap_err();
        assert!(matches!(err, Error::Truncated));
    }
}
